src/plugins/MediaManager.js

```diff
@@ -0,0 +1,341 @@
+     * @param {DeviceRenderer} instance root instance
+     * @param {boolean} videoWithMicrophone flag indicating wether the video must always be played with the microphone. Defaults to the microphone flag in the instance options
+     * @param {number} videoWidth Maximum video width. Defaults to 1280
+     * @param {number} videoHeight Maximum video height. Defaults to 720
+     */
+    constructor(instance, videoWithMicrophone = instance.options.microphone, videoWidth = 1280, videoHeight = 720) {
+        this.instance = instance;
+        this.instance.mediaManager = this;
+
+        if (!navigator.mediaDevices || !this.setupPermissions()) {
+            log.error('MediaDevices API unsupported: camera and microphone won\'t be available.');
+            this.instance.mediaEventsEnabled = false;
+            return;
+        }
+        this.instance.mediaEventsEnabled = true;
+
+        this.videoStreaming = false;
+        this.audioStreaming = false;
+
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoWithMicrophone = videoWithMicrophone;
+    }
+
+    /**
+     * Tries to get the current microphone and camera permission status. If it can, binds listeners to their change
+     * @returns {boolean} True if it could find the permissions object
+     */
+    setupPermissions() {
+        if (!navigator.permissions) {
+            return false;
+        }
+
+        navigator.permissions.query({name: 'microphone'})
+            .then((permissionObj) => {
+                log.debug(`microphone ${permissionObj.state}`);
+                permissionObj.addEventListener('change', this.onMicrophonePermissionChange.bind(this));
+            })
+            .catch((error) => {
+                log.debug('Can\'t get microphone permission object', error);
+            });
+
+        navigator.permissions.query({name: 'camera'})
+            .then((permissionObj) => {
+                log.debug(`camera ${permissionObj.state}`);
+                permissionObj.addEventListener('change', this.onCameraPermissionChange.bind(this));
+            })
+            .catch((error) => {
+                log.debug('Can\'t get camera permission object', error);
+            });
+
+        return true;
+    }
+
+    /**
+     * Toggle local video forwarding.
+     * Redirect the client webcam video stream to the instance.
+     */
+    toggleVideoStreaming() {
+        if (!this.videoStreaming) {
+            this.startVideoStreaming();
+        } else {
+            this.stopVideoStreaming();
+        }
+    }
+
+    /**
+     * Toggle local audio forwarding.
+     * Redirect the client microphone audio stream to the instance.
+     */
+    toggleAudioStreaming() {
+        if (!this.audioStreaming) {
+            this.startAudioStreaming();
+        } else {
+            this.stopAudioStreaming();
+        }
+    }
+
+    /**
+     * Initialize and start client webcam video stream.
+     */
+    startVideoStreaming() {
```

Comment:
then method is oldschool and can become unreadable as it grow. startVideoStreaming won't return a promise, so if you want to wait for startVideoStreaming in the future you can't without modify startVideoStreaming

i prefere this way

```
async startVideoStreaming() {
     if (!navigator.mediaDevices) {
         return;
     }
     try{
        const mediaStream = navigator.mediaDevices.getUserMedia({
            audio: this.videoWithMicrophone,
            video: {
                width: this.videoWidth,
                height: this.videoHeight,
            },
        })
        log.debug('Client video stream ready');
        this.videoStreaming = true;
        this.localVideoStream = mediaStream;
        this.addVideoStream(mediaStream);
      } catch((err)=> {
        this.onVideoStreamError(err);
      })           
}
```

Reply:
It is really interesting to read that .then() is old school and you propose a more conventional way of writing code. I really don't understand JS dev :)

Reply:
I was also surprised at this but indeed, the browser support versions seems to indicate that async/await + try/catch is much newer than promises

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#browser_compatibility